v20.1/build-a-spring-app-with-cockroachdb-jdbc.md

```diff
@@ -0,0 +1,709 @@
+
+`setTransactionAttributes` sets the transaction attributes in the form of advice. Spring supports [several different annotations to declare advice](https://docs.spring.io/spring/docs/current/spring-framework-reference/core.html#aop-advice). The [`@Around` annotation](https://docs.spring.io/spring/docs/current/spring-framework-reference/core.html#aop-ataspectj-around-advice) declares `setTransactionAttributes` as the advice to execute at a `anyTransactionBoundaryOperation(transactional)` join point.
+
+On verifying that the transaction is active (using `TransactionSynchronizationManager.isActualTransactionActive()`), the advice [sets some session variables](set-vars.html) using methods of the [`JdbcTemplate`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/jdbc/core/JdbcTemplate.html) object declared at the top of the `TransactionHintsAspect` class definition. These session variables (`application_name`, `statement_timeout`, and `transaction_read_only`) set [the application name](connection-parameters.html#additional-connection-parameters) for the query to "`roach-data`", the time allowed for the statement to execute before timing out to 1000 milliseconds (i.e., 1 second), and the [transaction access mode](set-transaction.html#parameters) as either `READ ONLY` or `READ WRITE`.
+
+#### Transaction retries
+
+To [avoid transaction contention](performance-best-practices-overview.html#understanding-and-avoiding-transaction-contention), we highly recommend writing [client-side transaction retry logic](transactions.html#client-side-intervention) into applications written on CockroachDB. In this application, transaction retry logic is written into the methods of the `RetryableTransactionAspect` class, which is declared as an aspect with the `@Aspect` annotation. The `@Order` annotation is passed `Ordered.LOWEST_PRECEDENCE-2`, a level of precedence above the primary transaction advisor. This indicates that this advisor must run outside the context of a transaction. Here are the contents of [`RetryableTransactionAspect.java`](https://github.com/cockroachlabs/roach-data/blob/master/roach-data-jdbc/src/main/java/io/roach/data/RetryableTransactionAspect.java):
```

Comment:
I think retries are a way to _handle_ aborted transactions due to serialization (transient) errors rather than a method for _avoiding_ them. Either through savepoints or retrying local transactions with exponential backoff. At least that's better than propagating transient errors all the way up to the stack to clients/user agents. Ultimately, contention avoidance by schema design and/or reviewing the interleaving of reads/writes is the best option, but not always feasible.

Reply:
Updated.